include/oneapi/tbb/detail/_machine.h

```diff
@@ -79,22 +79,20 @@ using std::this_thread::yield;
 // atomic_fence implementation
 //--------------------------------------------------------------------------------------------------
 
-#if _MSC_VER && (__TBB_x86_64 || __TBB_x86_32)
-#pragma intrinsic(_mm_mfence)
-#endif
-
 static inline void atomic_fence(std::memory_order order) {
-#if _MSC_VER && (__TBB_x86_64 || __TBB_x86_32)
```

Comment:
Microsoft Compiler used to generate [strange code](https://godbolt.org/z/q9zqn3MjG) that allocates two registers and slowdowns nearby code. For our use cases `mfence` is not big deal, rather than broken optimization due to inefficient register allocation. We still use old Microsoft Compiler for the releases and I would remain the workaround as is. 

GCC 11 generates really good code: https://godbolt.org/z/1eobjq5PT, so the proposed patch is going to be outdated.  So, it seems unnecessary for modern GCC and for official releases (we use Intel Compiler on Linux that does not bother itself and generates `mfence`).

If you think that the patch is useful for your use case, I would refactor out `order` argument because we use only `memory_order_seq_cst` and limit the workaround to GCC version prior to 11.

Reply:
> Microsoft Compiler used to generate strange code that allocates two registers and slowdowns nearby code.

The godbolt output you linked shows that MSVC does basically the same - replaces `mfence` with `lock cmpxchg`. Yes, `cmpxchg` is not the best replacement, but I would still expect it to be far better than `mfence`, even with two clobbered registers. See [here](https://uops.info/table.html?search=lock%20cmpxchg&cb_lat=on&cb_tp=on&cb_uops=on&cb_ports=on&cb_SKL=on&cb_ICL=on&cb_ZEN3=on&cb_measurements=on&cb_doc=on&cb_base=on&cb_sse=on) and [here](https://uops.info/table.html?search=mfence&cb_lat=on&cb_tp=on&cb_uops=on&cb_ports=on&cb_SKL=on&cb_ICL=on&cb_ZEN3=on&cb_measurements=on&cb_doc=on&cb_base=on&cb_sse=on) for performance numbers. The bottom line is `mfence` is very slow (about half the throughput), after all security mitigations, so replacing it with a `lock`-prefixed operation is a win.

I can leave the `mfence` in place if you insist, I just don't see the point.

> GCC 11 generates really good code, so the proposed patch is going to be outdated.

Yes, gcc 11 have optimized the thread fence intrinsics. It uses `lock or [rsp], 0` as a replacement though, which might be less efficient than the asm statement I proposed because:

- `or` instruction affects flags,
- `[rsp]` is not necessarily the location of the latest accessed data on the stack, as the compiler tends to allocate the stack frame beforehand. This means that `[rsp]` is less likely to be in the cache than a variable that is inside the stack frame, near the latest allocated variable in the preceding code.

I cannot say how much these nuances affect performance, probably not by much. I can limit the asm code to gcc versions prior to 11 if you prefer.

> If you think that the patch is useful for your use case, I would refactor out `order` argument

Sure, I can do that, though I'd prefer to remove `order` in a separate commit.

Reply:
I see the point. I checked the code, we already eliminated `atomic_fence` from the hottest paths, so even inefficient register allocation does not seem to be the issue. Hence, let's go with your solution because it is indifferent for Intel-based CPU.

> I can limit the asm code to gcc versions prior to 11 if you prefer, though I see no harm in leaving it enabled unconditionally. 

The main harm is two versions of the code that can become different over the time and require multiple configurations to test them. Moreover, the assembler limits compiler capabilities to optimize nearby code and generate better code for future CPUs (what if someone invents one more good instruction?). In addition, memory model is really complex topic and it is better to have implementation as simple as possible to avoid unforeseen issues and apply such tools like Thread Sanitizer to check the correctness but Thread Sanitizer does not understand assembler. Unfortunately, it does not understand standalone barriers but there are some attempts (mentioned in https://github.com/google/sanitizers/issues/1415) to add them. Therefore, I prefer to keep workarounds limited and rely on generic coded as well tested and reasonably performant solutions.

> Sure, I can do that, though I'd prefer to remove order in a separate commit.

Up to you, so you can consider not to create branches for other than `memory_order_seq_cst` memory order.

Reply:
Done.